### PR TITLE
feat: restructure, typescript with nodenext-support, adding isElasticsearchClient-helper

### PR DESCRIPTION
**Pull request by Cyberlane**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by Cyberlane** [comment]
@mcollina - no problem at all. I won't have a chance to do it before Monday I am afraid, but I will add it.

**Comment by Cyberlane** [comment]
@mcollina - sorry for the delay.

So I missed one thing with my PR, I forgot to update the package.json to have the `types` field set, I will fix that shortly.

As for your unit test request, I am not sure what you would like to be tested? The file I submitted is a type definition file to your vanilla node package, it doesn't directly affect your code and simply informs TypeScript what the structure of your code is.

However, if you prefer, I could rewrite your package into TypeScript, along with your existing unit tests, update your builds to generate vanilla JS in a dist folder that gets published. This would remove the need for the type definition file as it would be generated at build time.

Let me know how you would like me to proceed 😄

**Comment by kibertoad** [comment]
@Cyberlane We do not write fastify packages in TypeScript.
See https://github.com/fastify/fastify-postgres/blob/master/test/types/query.test-d.ts for an example of type tests

**Comment by Cyberlane** [comment]
@kibertoad thanks for the clarification, I will hook that up in the morning

**Comment by Cyberlane** [comment]
I wrote some basic tests, but it has revealed some unfortunate issues.

The `Client` type from `@elastic/elasticsearch` is in fact a class, which which means union it with a `Record<string, Client>` does not work for the namespaces.

Basically this:

```ts
import fastify from 'fastify';
import type { Client } from '@elastic/elasticsearch';
import { expectType } from 'tsd';

import fastifyElasticsearch from './index';

const app = fastify();
app.register(fastifyElasticsearch, {
  node: 'http://localhost:9200',
  namespace: 'cluster',
});

app.after(() => {
  expectType<Client>(app.elastic.cluster); // This is not happy about the type
});
```

However... if I replace `Client` with my own type, it works just fine. The issue is that it's not ideal to assign your own type, since there are [so many public properties](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/api-reference.html) and methods, which might also change.

I would like some feedback and thoughts on the preferred way to move forward. If you want your own type, I can do that, but it will be quite a bit to maintain down the road, but if you can think of a better solution please let me know.

**Comment by Uzlopak** [comment]
I have made a proposal to fix the issue:

https://github.com/Cyberlane/fastify-elasticsearch/pull/1/files

Created a typeguard which check if the path is a elastcsearchclient or not.

**Comment by Cyberlane** [comment]
@Uzlopak that's actually a fantastic way around this, thank you!

I've accepted your PR and merged it, I will fix the merge conflicts on this PR in a little bit :)

File: index.js [hunk 1]
```diff
@@ -10,6 +10,12 @@ async function fastifyElasticsearch (fastify, options) {
 
   const client = options.client || new Client(options)
 
+  if (!fastify.isElasticsearchClient) {
+    fastify.decorate('isElasticsearchClient', function isElasticsearchClient (value) {
```
Comment (mcollina) [inline index.js:14]: Why is this needed?

**Review by mcollina — CHANGES_REQUESTED** [review]

File: index.js (hunk 1 above)
Comment (Uzlopak) [reply index.js:14]: If the question is why it is decorated to the fastify.instance:
The reason, why I added that function/typeguard was, that I did not wanted to enforce the need to require the plugin again just to import isElasticsearchClient to use it for typescript purposes. 
Also the fastify-instance scope felt correct, as the elastic search instance is imho connected on fastify start. 

If the questoin is, why do we have isElasticsearchClient:
In Typescript you need to narrow down the type as the type is actually ambigous. And with isElasticseachClient we can use it to determine if we use namespaces or not. So this is more like a typescript helper method.

Comment (Uzlopak) [reply index.js:14]: If the question is why it is decorated to the fastify.instance:
The reason, why I added that function/typeguard was, that I did not wanted to enforce the need to require the plugin again just to import isElasticsearchClient to use it for typescript purposes. 
Also the fastify-instance scope felt correct, as the elastic search instance is imho connected on fastify start. 

If the questoin is, why do we have isElasticsearchClient:
In Typescript you need to narrow down the type as the type is actually ambigous. And with isElasticseachClient we can use it to determine if we use namespaces or not. So this is more like a typescript helper method.

**Comment by Cyberlane** [comment]
@Uzlopak another option could be to embed the helper into the test file, and we could do something like this:

```ts
import fastifyElasticsearch from '..';
import Fastify from 'fastify';
import { Client } from '@elastic/elasticsearch';
import { expectError, expectNotType, expectType } from 'tsd';

const fastify = Fastify()

type ElasticInstance = Client & Record<string, Client>;

fastify.register(fastifyElasticsearch, { node: 'http:localhost:9200' })

const isElasticClient = (value: any): boolean => {
  if (value instanceof Client) {
    return true;
  }
  throw new Error('Not instance of ElasticSearch Client');
}

expectType<ElasticInstance>(fastify.elastic)
expectNotType<Client>(fastify.elastic)
expectNotType<ElasticInstance>(fastify.elastic.abc)
expectType<Client>(fastify.elastic.abc)

expectType<boolean>(isElasticClient(fastify.elastic))
expectError<Error>(isElasticClient(fastify.elastic.abc))

fastify.register(fastifyElasticsearch, { node: 'http://localhost:9200', namespace: 'abc' })

expectType<boolean>(isElasticClient(fastify.elastic.abc))

```

**Comment by Uzlopak** [comment]
Well first of all expectError from tsd is not about throwing an Error but expecting a typescript error. 

In my humble opinion I think isElasticsearchClient is just a simple typescript-typeguard and it should be in the implementation only a function which returns true or false. Just to uphold the "contract" between typescrpit and javascript. 

A pure javascript developer will probably not use isElasticsearchClient as the developer should know if namespaces are used or not. But for the typescript dev it is also about typings. And for this one edge case the developer should use isElasticsearchClient. And if the developer wants to throw than it can be easily added. But if we throw always if it is not a elasticsearch-Client than we are not providing a typeguard and if the dev wants just to ensure it is a Client the function has to be wrapped in a try-catch-block.

So I would actually try to keep it very very simple.

**Comment by Uzlopak** [comment]
@mcollina PTAL

I modified the PR and now isElasticsearchClient is a helper method provided by the package itself and only used for typescript type narrowing. It doesnt "taint" the fastify-instance with an additional decoration. 

Also nodenext compatibility and minor restructuring.

**Review by Eomm — APPROVED** [review]

**Review by mcollina — APPROVED** [review]
lgtm
